src/TextInputWithTokens.tsx

```diff
@@ -0,0 +1,247 @@
+import React, { ComponentPropsWithoutRef, FocusEventHandler, KeyboardEventHandler, useEffect, useRef, useState } from 'react'
+import {omit} from '@styled-system/props'
+import styled from 'styled-components'
+import { FocusKeys } from './behaviors/focusZone'
+import { useCombinedRefs } from './hooks/useCombinedRefs'
+import { useFocusZone } from './hooks/useFocusZone'
+import {ComponentProps, MandateProps} from './utils/types'
+import Token, { TokenProps } from './Token/Token'
+import TokenLabel, { TokenLabelProps } from './Token/TokenLabel'
+import TokenProfile, { TokenProfileProps } from './Token/TokenProfile'
+import { TokenSizeKeys } from './Token/TokenBase'
+import TextInput, { TextInputProps } from './TextInput'
+import { useProvidedRefOrCreate } from './hooks'
+import UnstyledTextInput from './_UnstyledTextInput'
+
+const InputWrapper = styled.div`
+  order: 1;
+  flex-grow: 1;
+`
+
+type AnyTokenProps = Partial<TokenProps & TokenLabelProps & TokenProfileProps>
+type TokenDatum = MandateProps<AnyTokenProps, 'id' | 'text'>
+type TextInputWithTokensInternalProps = {
+  /**
+   * The array of tokens to render
+   */
+  tokens: TokenDatum[]
```

Comment:
I made a quick prototype of how we could define the type of `tokens` based on the type of `tokenComponent`: https://codesandbox.io/s/frosty-swanson-y8vyq?file=/src/App.tsx